url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/com/mxgraph/canvas/mxVmlCanvas.java

```diff
@@ -0,0 +1,604 @@
+					+ (int) (0.25 * h);
+
+			elem.setAttribute("path", points + " x e");
+		}
+		else if (shape.equals(mxConstants.SHAPE_ACTOR))
+		{
+			elem = document.createElement("v:shape");
+			elem.setAttribute("coordsize", w + " " + h);
+
+			double width3 = w / 3;
+			String points = "m 0 " + (h) + " C 0 " + (3 * h / 5)
+					+ " 0 " + (2 * h / 5) + " " + (w / 2) + " "
+					+ (2 * h / 5) + " c " + (int) (w / 2 - width3) + " "
+					+ (2 * h / 5) + " " + (int) (w / 2 - width3) + " 0 "
+					+ (w / 2) + " 0 c " + (int) (w / 2 + width3) + " 0 "
+					+ (int) (w / 2 + width3) + " " + (2 * h / 5) + " "
+					+ (w / 2) + " " + (2 * h / 5) + " c "
+					+ (w) + " " + (2 * h / 5) + " " + (w)
+					+ " " + (3 * h / 5) + " " + (w) + " "
+					+ (h);
+
+			elem.setAttribute("path", points + " x e");
+		}
+		else if (shape.equals(mxConstants.SHAPE_CYLINDER))
+		{
+			elem = document.createElement("v:shape");
+			elem.setAttribute("coordsize", w + " " + h);
+
+			double dy = Math.min(40, Math.floor(h / 5));
+			String points = "m 0 " + (int) (dy) + " C 0 " + (int) (dy / 3)
+					+ " " + (w) + " " + (int) (dy / 3) + " " + (w)
+					+ " " + (int) (dy) + " L " + (w) + " "
+					+ (int) (h - dy) + " C " + (w) + " "
+					+ (int) (h + dy / 3) + " 0 " + (int) (h + dy / 3) + " 0 "
+					+ (int) (h - dy) + " x e" + " m 0 " + (int) (dy) + " C 0 "
+					+ (int) (2 * dy) + " " + (w) + " " + (int) (2 * dy)
+					+ " " + (w) + " " + (int) (dy);
+
+			elem.setAttribute("path", points + " e");
+		}
+		else
+		{
+			if (mxUtils.isTrue(style, mxConstants.STYLE_ROUNDED, false))
+			{
+				elem = document.createElement("v:roundrect");
+				elem.setAttribute("arcsize",
+						(mxConstants.RECTANGLE_ROUNDING_FACTOR * 100) + "%");
+			}
+			else
+			{
+				elem = document.createElement("v:rect");
+			}
+		}
+
+		String s = "position:absolute;left:" + String.valueOf(x) + "px;top:"
+				+ String.valueOf(y) + "px;width:" + String.valueOf(w)
+				+ "px;height:" + String.valueOf(h) + "px;";
+
+		// Applies rotation
+		double rotation = mxUtils.getDouble(style, mxConstants.STYLE_ROTATION);
+
+		if (rotation != 0)
+		{
+			s += "rotation:" + rotation + ";";
+		}
+
+		elem.setAttribute("style", s);
+
+		// Adds the shadow element
+		if (mxUtils.isTrue(style, mxConstants.STYLE_SHADOW, false)
+				&& fillColor != null)
+		{
+			Element shadow = document.createElement("v:shadow");
+			shadow.setAttribute("on", "true");
+			shadow.setAttribute("color", mxConstants.W3C_SHADOWCOLOR);
+			elem.appendChild(shadow);
+		}
+
+		float opacity = mxUtils.getFloat(style, mxConstants.STYLE_OPACITY, 100);
+
+		// Applies opacity to fill
+		if (fillColor != null)
+		{
+			Element fill = document.createElement("v:fill");
+			fill.setAttribute("color", fillColor);
+
+			if (opacity != 100)
+			{
+				fill.setAttribute("opacity", String.valueOf(opacity / 100));
+			}
+
+			elem.appendChild(fill);
+		}
+		else
+		{
+			elem.setAttribute("filled", "false");
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 487:502, 505:520.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.